### PR TITLE
Add a markdown with high level kubernetes metadata enrichment explanation

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
- WHAT: Add a markdown which explains in high level how kubernetes metadata enrichment works in metrics modules
- WHY:  the new implementation of https://github.com/elastic/beats/pull/37332 needs a high level explanation

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+## Kubernetes Metadata enrichment
+
+[metadata diag](../_meta/images/enrichers.png)
```
Comment (constanca-m) [inline metricbeat/module/kubernetes/util/enrichers.md:3]: I can't see the image in the preview, can you see it @MichaelKatsoulis ?

Comment (constanca-m) [reply metricbeat/module/kubernetes/util/enrichers.md:3]: I also don't understand what the image is trying to say... Is it just an example using 3 metricsets on what they are doing?

**Comment by constanca-m** [comment]
Should this part: https://github.com/elastic/beats/blob/8efb36685ea9213a7f0f9ad583698cdf85394d3e/metricbeat/module/kubernetes/util/kubernetes.go#L911-L924, be mentioned somehow or do you think it's too specific?

File: metricbeat/module/kubernetes/util/enrichers.md (hunk 1 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/kubernetes/util/enrichers.md:3]: It is supposed to show an example of different metricsets having their own metadata enricher but sharing same watchers which use the k8s API to watch for a resource. 
I am open to all kind of suggestions !!

Comment (constanca-m) [reply metricbeat/module/kubernetes/util/enrichers.md:3]: Ok! Maybe you can just write a sentence below the image saying that?

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+The metadata enrichment process involves associating contextual information, such as Kubernetes metadata (e.g., labels, annotations, resource names), with metrics and events collected by Elastic Agent and Beats in Kubernetes environments. This process enhances the understanding and analysis of collected data by providing additional context.
+
+### Key Components:
+
+1. **Metricsets:**
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:9]: To differentiate with datasets. Just writ something same term is called datasets in elastic-agent.
https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+   - Metricsets are responsible for collecting metrics and events from various sources within Kubernetes, such as kubelet and kube-state-metrics.
+
+2. **Enrichers:**
+   - Enrichers are components responsible for enriching collected data with Kubernetes metadata. Each metricset is associated with its enricher, which handles the metadata enrichment process.
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:13]:
```suggestion
   - Enrichers are components responsible for enriching collected data with Kubernetes metadata. Each metricset is associated with its enricher, which handles the metadata enrichment process.
```

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+   - Metadata generators are responsible for generating metadata associated with Kubernetes resources. These generators are utilized by enrichers to collect relevant metadata. Each enricher has one metadata generator.
+
+### Metadata Generation Process:
+
+1. **Initialization:**
+   - Metricsets are initialized with their respective enrichers during startup. Enrichers are responsible for managing the metadata enrichment process for their associated metricsets.
+
+2. **Watcher Creation:**
+   - Watchers are created to monitor Kubernetes resources relevant to the metricset's data collection requirements. For example pod metricset triggers the creation of watcher for pods, nodes and namespaces.
+   - Multiple enrichers are associated with one watcher. For example a pod watcher is associated with pod, state_pod, container and state_container metricsets and their enrichers.
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:28]: I think that this line should go above and then `- Watchers are created ...`

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+   - Metricsets are initialized with their respective enrichers during startup. Enrichers are responsible for managing the metadata enrichment process for their associated metricsets.
+
+2. **Watcher Creation:**
+   - Watchers are created to monitor Kubernetes resources relevant to the metricset's data collection requirements. For example pod metricset triggers the creation of watcher for pods, nodes and namespaces.
+   - Multiple enrichers are associated with one watcher. For example a pod watcher is associated with pod, state_pod, container and state_container metricsets and their enrichers. 
+
+3. **Metadata Generation:**
+   - When a watcher detects a change in a monitored resource (e.g., a new pod creation or a label update), it triggers the associated enrichers' metadata generation process.
+
+4. **Enrichment:**
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:33]:
```suggestion
4. **Enrichment Generation Process:**
```

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+   - The enricher collects relevant metadata from Kubernetes API objects corresponding to the detected changes. This metadata includes information like labels, annotations, resource names, etc.
+
+5. **Association with Events:**
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:36]: The association is actually the new enriched event tha happens on beats/ agent level. 
I think that we should clarify that a new enhanced event is sent to Elasticsearch

**Comment by gizas** [comment]
I think that the image is a little confusing. Maybe to put arrows to identify direction?
How about removing it and adding a table with what watchers are being created per metricset?

**Comment by MichaelKatsoulis** [comment]
> Should this part:
> 
> https://github.com/elastic/beats/blob/8efb36685ea9213a7f0f9ad583698cdf85394d3e/metricbeat/module/kubernetes/util/kubernetes.go#L911-L924
> 
> , be mentioned somehow or do you think it's too specific?

Yes, I will add more info.

**Comment by MichaelKatsoulis** [comment]
@constanca-m and @gizas I answered all your review comments. Can you give it another go?

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+   - Watchers are created to monitor Kubernetes resources relevant to the metricset's data collection requirements. For example pod metricset triggers the creation of watcher for pods, nodes and namespaces.
+
+3. **Metadata Generation:**
+   - When a watcher detects a change in a monitored resource (e.g., a new pod creation or a label update), it triggers the associated enrichers' metadata generation process.
+
+4. **Enrichment Generation Process:**
+   - The enricher collects relevant metadata from Kubernetes API objects corresponding to the detected changes. This metadata includes information like labels, annotations, resource names, etc.
+
+5. **Association with Events:**
+   - The collected metadata is then associated with the metricset's events. This association enriches the events with contextual information, providing deeper insights into the collected data. The enriched events generated from beats/agent are then sent to the configured output (e.g. Elasticsearch).
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:35]:
```suggestion
   - The collected metadata are associated with the metricset's events. This association process enriches the events with contextual information, providing deeper insights into the collected data. The enriched events generated from beats/agent are sent to the configured output (e.g. Elasticsearch).
```

File: metricbeat/module/kubernetes/util/enrichers.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
+## Kubernetes Metadata enrichment
+
+The metadata enrichment process involves associating contextual information, such as Kubernetes metadata (e.g., labels, annotations, resource names), with metrics and events collected by Elastic Agent and Beats in Kubernetes environments. This process enhances the understanding and analysis of collected data by providing additional context.
+
+### Key Components:
+
+1. **Metricsets/Datasets:**
+   - Metricsets/Datasets are responsible for collecting metrics and events from various sources within Kubernetes, such as kubelet and kube-state-metrics.
```
Comment (gizas) [inline metricbeat/module/kubernetes/util/enrichers.md:8]: Still dont explain why we have two terms. Users dont understand why this changes in beats vs agents. I think we should make this clear

**Review by gizas — APPROVED** [review]

**Review by constanca-m — APPROVED** [review]

File: metricbeat/module/kubernetes/util/enrichers.md (hunk 8 above)
Comment (MichaelKatsoulis) [reply metricbeat/module/kubernetes/util/enrichers.md:8]: @gizas  This markdown is in beats repo, under kubernetes module. There, only metricsets exist. Personally I would remove the `dataset` as it is confusing. What kind of sentence do you want me to add?

Comment (gizas) [reply metricbeat/module/kubernetes/util/enrichers.md:8]: Missed that we are under beats. I had in my mind that this could be a more general doc. 
So ok you can remove datasets overall
